tools/istio-iptables/pkg/builder/rules_builder.go

```diff
@@ -0,0 +1,52 @@
+package builder
+
+type Rule struct {
+	chain  *string
+	table  *string
+	params []string
+}
+
+type Rules struct {
+	rules   []*Rule
+	rulesv6 []*Rule
+}
+
+type Builder interface {
```

Comment:
I've been thinking that the code that adds rules doesn't care about the `iptable-restore` input, and vice-versa the code that needs to get the strings for `iptables-restore` shouldn't have to add rules.
So there are 2 distinct responsibilities here. I'd split this interface into 2, one for each responsibility. This will make it easier to mock each interface in unit tests.
Your concrete implementation below can implement both interfaces. No need to change that.

```
type IptablesProducer interface {
	AddV4Rule(chain string, table string, params ...string) IptablesProducer
	AddV6Rule(chain string, table string, params ...string) IptablesProducer
}

type IptablesConsumer {
	BuildV4Rules() []string
	BuildV6Rules() []string
	BuildIptablesRestore() string
	BuildIp6tablesRestore() string
}

type IptablesBuilder interface {
	IptablesProducer
	IptablesConsumer
}
```

This way you can inject an `IptablesProducer` to one part of the code, and an `IptablesConsumer` to another.

I'm not set on the producer/consumer naming. I'm open to better names for those.

Reply:
hmm.. let me think about this